File: 76.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:

        mp = {}
        ans_mp = {}
        min_len = 1000000
        ans_string = ""
        for c in t:
            ans_mp[c] = 1 + ans_mp.get(c, 0)

        i = 0
        j = 0

        for c in s:
            mp[c] = 1 + mp.get(c, 0)
            suc = 1
            for k in ans_mp.keys():
                if mp.get(k, 0) < ans_mp.get(k, 0):
                    suc = 0
                    break
            if suc == 1:

                while j <= i and suc:

                    # can take this out to make efficient
                    if min_len > (i-j+1):
                        ans_string = ""
                        min_len = (i-j+1)
                        for k in range(j, i+1):
                            ans_string += s[k]

                    mp[s[j]] -= 1
                    j += 1

                    if mp.get(s[j-1], 0) < ans_mp.get(s[j-1], 0):
                        suc = 0
                        break

            i += 1

        return ans_string


if __name__ == '__main__':

    s = input()
    t = input()

    sol = Solution()
    print(sol.minWindow(s, t))

File: 76.cpp
/*Author : Manraj Singh */
#include <bits/stdc++.h>
using namespace std;

// Problem :
/* Solution : 	 */

class Solution
{
public:
    string minWindow(string s, string t)
    {

        string ans = "";
        int min_len = INT_MAX , i = 0 , j = 0;
        unordered_map<char , int > mp , ans_mp;
        for (int i = 0; i < t.size(); i++)
        {
            ans_mp[t[i]]++;
        }
        
        while (i < s.size())
        {
            mp[s[i]]++;
            int suc = 1;
            for(auto ss : ans_mp){

                char c = ss.first;
                if ( mp[c] < ans_mp[c]){

                    suc = 0;
                    break;

                }
            }

            // cout<<"success = "<<suc<<endl;

            if ( suc == 1){

                int min_i = i, min_j=j;
                while (suc && j <= i)
                {
                    
                    mp[s[j]]--;
                    min_j = j;
                    j++;
                    if (mp[s[j-1]] < ans_mp[s[j-1]]){
                        suc = 0;
                        break;
                    }

                }

                // cout<<"i = "<<min_i<<"min j = "<<min_j<<endl;

                if ( min_len > ( min_i - min_j + 1)){

                    min_len = (min_i-min_j + 1);
                    ans = "";
                    for (int k = min_j; k <= min_i; k++)
                    {
                        ans += s[k];
                    }
                    
                }
                // cout<<"ans = "<<ans<<endl;
                
            }

            if ( j > i ){
                j = i;
            }
            
            // cout<<i<<" "<<j<<endl;

            i++;

            // cout<<"ans= "<<ans<<endl;
        }
        

        return ans;
    }
}

;

int main()
{

    ios::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    string a, b;
    cin >> a >> b;

    Solution s;
    cout<<s.minWindow(a, b);

    return 0;
}